File: src/models/trade.ts
/**
 * Trade record - tracks individual buy/sell/rotation transactions
 * Links to spec Section 7 (Trade Execution)
 */

export interface Trade {
  readonly id: string;
  readonly timestamp: Date;
  readonly ticker: string;
  readonly type: 'BUY' | 'SELL' | 'ROTATION_IN' | 'ROTATION_OUT';
  readonly price: number;
  readonly shares: number;
  readonly totalValue: number;
  readonly score?: number;
  readonly reason?: string;
}

export class TradeRecord {
  private trades: Trade[] = [];
  private nextId = 0;

  /**
   * Record a new trade
   */
  recordTrade(
    ticker: string,
    type: Trade['type'],
    price: number,
    shares: number,
    score?: number,
    reason?: string
  ): Trade {
    const trade: Trade = {
      id: `trade_${this.nextId++}`,
      timestamp: new Date(),
      ticker,
      type,
      price,
      shares,
      totalValue: price * shares,
      score,
      reason,
    };

    this.trades.push(trade);
    return trade;
  }

  /**
   * Get all trades
   */
  getTrades(): readonly Trade[] {
    return [...this.trades];
  }

  /**
   * Get trades for a specific ticker
   */
  getTradesForTicker(ticker: string): Trade[] {
    return this.trades.filter((t) => t.ticker === ticker);
  }

  /**
   * Get trades of a specific type
   */
  getTradesOfType(type: Trade['type']): Trade[] {
    return this.trades.filter((t) => t.type === type);
  }

  /**
   * Get most recent trades (for diagnostics)
   */
  getRecentTrades(count: number): Trade[] {
    return [...this.trades].reverse().slice(0, count);
  }

  /**
   * Clear trade history (for backtesting)
   */
  clear(): void {
    this.trades = [];
    this.nextId = 0;
  }
}

File: src/models/trade.py
"""Trade representation."""

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum


class TradeType(Enum):
    """Trade type enumeration."""
    BUY = "BUY"
    SELL = "SELL"
    ROTATION = "ROTATION"


@dataclass
class Trade:
    """Represents a single trade execution."""
    
    ticker: str
    trade_type: TradeType
    shares: int
    price: float
    timestamp: datetime = field(default_factory=datetime.now)
    score: float = 0.0
    reason: str = ""
    commission: float = 0.0
    
    @property
    def trade_value(self) -> float:
        """Calculate total trade value."""
        return self.shares * self.price
    
    @property
    def total_cost(self) -> float:
        """Calculate total cost including commission."""
        return self.trade_value + (self.commission if self.trade_type == TradeType.BUY else -self.commission)
    
    def to_dict(self) -> dict:
        """Convert to dictionary representation."""
        return {
            "ticker": self.ticker,
            "type": self.trade_type.value,
            "shares": self.shares,
            "price": round(self.price, 2),
            "value": round(self.trade_value, 2),
            "commission": round(self.commission, 2),
            "total_cost": round(self.total_cost, 2),
            "score": round(self.score, 4),
            "reason": self.reason,
            "timestamp": self.timestamp.isoformat(),
        }
